arangod/Scheduler/Scheduler.cpp

```diff
@@ -208,28 +212,76 @@ Scheduler::~Scheduler() {
 }
 
 // do not pass callback by reference, might get deleted before execution
-void Scheduler::post(std::function<void()> const& callback) {
+void Scheduler::post(std::function<void()> const callback, bool isV8,
+                     uint64_t timeout) {
   incQueued();
 
   try {
+    if (isV8) {
+      ++_queuedV8;
+    }
+
     // capture without self, ioContext will not live longer than scheduler
-    _ioContext.get()->post([this, callback]() {
-      JobGuard guard(this);
-      guard.work();
+    _ioContext->post([this, callback, isV8, timeout]() {
+      try {
+        if (isV8 && _queuedV8 > _maxQueuedV8 &&
+            numWorking(getCounters()) >= static_cast<uint64_t>(_maxQueuedV8)) {
+          --_queuedV8;
+          decQueued();
 
-      decQueued();
+          auto t = timeout;
 
-      callback();
+          if (t == 0) {
+            t = 10;
+          } else {
+            t *= 2;
+
+            if (200 <= t) {
+              t = 200;
+            }
+          }
+
+          std::shared_ptr<asio_ns::deadline_timer> timer(
+              newDeadlineTimer(boost::posix_time::millisec(timeout)));
+          timer->async_wait(
+              [this, callback, isV8, t](const asio::error_code& error) {
+                if (error != asio::error::operation_aborted) {
+                  post(callback, isV8, t);
+                }
+              });
+
+          return;
+        }
+
+        decQueued();
+
+        callback();
+
+        if (isV8) {
+          --_queuedV8;
+        }
+      } catch (...) {
+        if (isV8) {
+          --_queuedV8;
+        }
+
+        throw;
+      }
     });
   } catch (...) {
     decQueued();
+
+    if (isV8) {
+      --_queuedV8;
+    }
+
     throw;
   }
 }
 
 // do not pass callback by reference, might get deleted before execution
 void Scheduler::post(asio_ns::io_context::strand& strand,
-                     std::function<void()> const& callback) {
+                     std::function<void()> const callback) {
```

Comment:
why are we copying the callback here? The comment mentions it, but I am wondering if the callback can actually be deleted from "outside".

Reply:
done